### PR TITLE
Fix #3657: Remove a number of java.net defects

### LINKED ISSUE #3657: javalib Inet6Address related technical debt
javalib has some Inet6Address related technical debt.  An incomplete list:

1) Class Inet6Address is missing the static method `getByAddress(String host, byte[] addr, NetworkInterface nif)`.

2) Class Inet6Address is missing the instance method  `getScopedInterface()`

3) In Class InetAddress the entire line `protected def getZoneIdent(): String = ""` should be deleted.  Looks like
    that is scaffolding left over from early Scala Native Inet6Address development.  A new `getScopedInterface()`
    can be used to get the same information more idiomatically.

Changing these around may  affect the constructor arguments for `Inet6Address`.

### PR DESCRIPTION
**Pull request by LeeTibbert**
Fix #3657

A number of  defects in javalib  java.net were removed. See referenced Issue for details.

In particular, `InterfaceAddress` and `InetAddress` instances returned by `NetworkInterface`
should handle the network interface index as they are handled in the JVM. A number of
other changes are internal and not user-facing.

Since `NetworkInterface` was not, for good reasons, ported to SN 0.4.n, this PR is __not__ a 
candidate for backporting.

### CODE REVIEW

File: javalib/src/main/scala/java/net/Inet6Address.scala
Comment (ekrich) [inline javalib/src/main/scala/java/net/Inet6Address.scala:164]: While you are in here, you might want to change this to a normal `while` loop with a `var`.

**Review by ekrich — APPROVED** [review]
This looks great. Of course I am not digging into the correctness. Just the one suggestion.

File: javalib/src/main/scala/java/net/Inet6Address.scala
Comment (LeeTibbert) [reply javalib/src/main/scala/java/net/Inet6Address.scala:164]: Thank you for the review (and encouragement).

I thought/think that `for()` with a range was optimized in Scala Native quite a long time ago,
perhaps 4 or more years.  Perhaps both ends of the range need to be constants for that to happen?

I will have to try to figure out how to prove to myself if such a loop is optimized.

Comment (ekrich) [reply javalib/src/main/scala/java/net/Inet6Address.scala:164]: I did find this page so maybe it is ok. I prefer the `while` although more prone to error. https://github.com/scala/bug/issues/1338
```scala
scala> 0 until 10
res0: scala.collection.immutable.Range = Range 0 until 10
```
Whether the Range construction is completely optimized out only byte code would tell.

Comment (LeeTibbert) [reply javalib/src/main/scala/java/net/Inet6Address.scala]: I went with the magic number constant 16. Passes on my build systems. Let's see if CI
likes it.  An Inet6Address is only created by javalib, never users. If the size of a
binary IPv6 address ever changes, it will be evident early in the development process:
things will break all over and in a lot more places that that loop.

Comment (LeeTibbert) [reply javalib/src/main/scala/java/net/Inet6Address.scala]: Eric,

Hard to measure, but your suggestion may have lead to at least a marginal improvement in
the speed of formatting an IP6Address. Those are used all over java.net unit-tests. Once again:
Thank you.

**Comment by LeeTibbert** [comment]
There are two failures which look like they have nothing to do with the essence of this PR.  I
am going to wait overnight to see if the pixies run the failing tests again.

The failures are both on macOS 12. One on Java 11 and the other on Java 21.  The Java 17 
test succeeds.   Tests on my private macOS 14.2.n system running either Java 8 or Java 21
succeed. 

Guess this is why we run CI.

**Comment by LeeTibbert** [comment]
@RustedBones 

If you have interest and opportunity, could you look at these changes and let me know what you think?
You may not be able to do a formal GitHub review, but should be able to create replies, such as this one.
Thank you. I understand and appreciate if Life demands you be elsewhere.

All this started from you concern about consistency of constructor arguments.

**Comment by LeeTibbert** [comment]
Reviewers,

I hope this Reply saves you some time.  Sorry for the complexity.  
You probably do not want to spend time to trace if `scope_id` a.k.a "network interface id" gets
passed cleanly from Java-to-C and back again from C-to-Java.  To my eyes, it does not.  I created
Issue #3672 to describe the details.  


I propose that this PR continue through the process, with this known defect in mind, and that I 
address it in a followup PR.  The defect occurs in the current mainline code, so no new defect is
introduced by this PR.

There are a number of disparate changes in this PR already, I would like to make the scope_id 
changes in a PR of its own, so reviewers and future maintainers can find them.

File: javalib/src/main/scala/java/net/Inet6Address.scala
Comment (WojciechMazur) [inline javalib/src/main/scala/java/net/Inet6Address.scala:49]: A good practice is to combine hashes by first multiply accumulated hashCode with some prime number to reduce chance of conflicts, eg. ` 
```
var res: Int = 1
res = 31 * res + ju.Arrays.hashCode(ipAddress) 
res = 31 * res + host.hashCode()
res = 31 * res + useScopeId.hashCode()
res = 31 * res + scopeId.hashCode()
res
```
see, `java.util.Arrays.hashCode` or  our implementation of it in `java.util.Arrays.hashCodeImpl`

**Review by WojciechMazur — COMMENTED** [review]
As always great job! Looks good to me, I have only 1 comment about calculating hashCodes

File: javalib/src/main/scala/java/net/Inet6Address.scala
Comment (LeeTibbert) [reply javalib/src/main/scala/java/net/Inet6Address.scala:49]: Wojciech,

Thank you for the review and, especially, the welcome short tutorial. I always learn things.

I reviewed the other uses of hashcode in `java.net` and will be updating those usages.

**Comment by LeeTibbert** [comment]
I believe this is ready for a second review.

The standard whine about failing tests...  Perhaps they will get automagically re-run over night.
Even though I am pretty sure the failures have nothing to do with the substance of this PR,
I never quite rest easy until all tests run successfully, especially with network code.

Both tests are on macOS.

The hard failing scala_next test  reports failed tests but the detailed log file gives, at least to me, no clue
as to why.

The scripted test has been going for 45+ minutes and looks like it is going to time out.

File: javalib/src/main/scala/java/net/Inet6Address.scala
Comment (RustedBones) [inline javalib/src/main/scala/java/net/Inet6Address.scala:145]: Maybe for my understanding, why `final` ? Since the is defined in an object this hasn't any effect right ?

File: javalib/src/main/scala/java/net/InetAddress.scala
Comment (RustedBones) [inline javalib/src/main/scala/java/net/InetAddress.scala:227]: `addrinfoToByteArray` already handles this. We can call the function and check the length of the returned array instead. This would avoid to test for IPv4 mapped address twice. WDYT ?

File: javalib/src/main/scala/java/net/Inet6Address.scala
Comment (WojciechMazur) [reply javalib/src/main/scala/java/net/Inet6Address.scala:145]: Making it final can make it a literal constant in compiler, but for that probably it's name should start with upper case. 
```suggestion
  private final val HexCharacters = "0123456789abcdef"
```

Comment (LeeTibbert) [reply javalib/src/main/scala/java/net/Inet6Address.scala]: RustedBones,

Thank you for looking at this, and in detail.

Yes, my intent in adding `final` was to increase the possibility that a little know & not obvious 
compiler optimization would happen.  I believe Scala itself has no concept of "const".

Wojciech,

Is the intent of making the initial character uppercase so that it can potentially be used in
pattern matching?  That convention for `final val` is new to me, times change and hopefully improve. 
Thank you for kindly picking up on my obsolete ways. 

The suggested change is easy enough to make here. With your consent, I can apply the convention
 i other files as I may touch them. Normally, I would  doing a sweep through the code base looking for
similar usages. I do not feel the benefit justifies the potential breakage, especially in files with which
I am not familiar.  Please advise.

File: javalib/src/main/scala/java/net/InetAddress.scala
Comment (LeeTibbert) [reply javalib/src/main/scala/java/net/InetAddress.scala:227]: re: // Scala JVM down-converts even when preferIPv6Addresses is "true"

I am not sure if I still believe that comment.  I will have to check it out.

I believe that JVM will __display_ an IPv4 mapped IPv6 address.  My reading 
since that comment was written makes me think that the JVM carries IPv4
mapped IPv6 addresses internally as IPv6 and not IPv4. 

I just do not remember if SN allows parsing IPv4 mapped IPv6 addresses, potentially with
an interface_index/scope_id on IPv4 only systems. My guess would be SN yes, JVM no.

On systems where the network stack is dual protocol IPv4/IPv6, the current internal is probably
invisible if the other end of the connection is also dual stack (and has chosen to listen on both
protocols, the default). When SN downconverts internally (i.e. early, just after parse) and the
remote end is listening on only IPv6, there might be a defect.

I am told, recently, that not all systems are "dual stack". Specifically I am told that FreeBSD and NetBSD
like to maintain independent IPv4 & IPv6 stacks (and can configure either or both off).  I believe,
but have not tested, that the current 'downconvert early' practice will break there. 

I wasn't doing anything in Q1, Q2, and possibly Q3 2024 anyway.

**Comment by LeeTibbert** [comment]
I believe this PR is ready for final review.

CI has three runs with errors.  I examined the log files and it looks to me
like it is a problem in mainline.  

Looks like today's PR #3674 has two similar failing JDK compliance tests.
Then again, a number of PRs submitted & merged in the past week 
managed to pass all tests. 

Perhaps I should be providing pennies for the eyes?

**Review by WojciechMazur — APPROVED** [review]

**Comment by LeeTibbert** [comment]
My thanks to RustedBones, Ekrich, and Wojciech for getting this merged.

**Pull request merged**
